tests/unit/test_expression_tree/test_variable.py

```diff
@@ -2,105 +2,95 @@
 # Tests for the Variable class
 #
 
-import unittest
+import pytest
 
 import numpy as np
 
 import pybamm
 import sympy
 
 
-class TestVariable(unittest.TestCase):
+class TestVariable:
     def test_variable_init(self):
         a = pybamm.Variable("a")
-        self.assertEqual(a.name, "a")
-        self.assertEqual(a.domain, [])
+        assert a.name == "a"
+        assert a.domain == []
         a = pybamm.Variable("a", domain=["test"])
-        self.assertEqual(a.domain[0], "test")
-        self.assertRaises(TypeError, pybamm.Variable("a", domain="test"))
-        self.assertEqual(a.scale, 1)
-        self.assertEqual(a.reference, 0)
+        assert a.domain[0] == "test"
+        pytest.raises(TypeError, match=pybamm.Variable("a", domain="test"))
```

Comment:
Could you take a look at this failure? – I noticed from the annotation, which was quite helpful. It looks like a `pytestify` bug, where, instead of converting `assertRaises` to a

```python
with pytest.raises(TypeError):
    pybamm.Variable("a", domain="test")
```

it was converted to a regular test where `match` expects a string, but it received a `pybamm.Variable` instance, which won't work.

Reply:
Also not sure if `TypeError` should be raised here anyways. See : https://github.com/pybamm-team/PyBaMM/blob/396741eec6076a83c3104aa9b73d2f7957201ab7/src/pybamm/type_definitions.py#L17
i.e. domain can be of type `string` or a list of it as well.

Reply:
```python
self.assertRaises(TypeError, pybamm.Variable("a", domain="test"))
```

works fine, as you mentioned, but

even the `unittest` equivalent of the `pytest.raises` context manager, i.e.,

```python
with self.assertRaises(TypeError):
    pybamm.Variable("a", domain="test")
```

fails with the same error not being raised.

I don't think this test was written correctly, since we should always use a context manager to catch errors. Either that, or we never enforced the type in the first place. For example, here is a short snippet of what happens if we enforce the type and raise an error:

```python
from typing import TypeAlias, Union
import unittest
import pytest


DomainType: TypeAlias = Union[list[str], str, None]


def foo(name: str, domain: DomainType = None) -> tuple[str, DomainType]:
    if isinstance(domain, int):
        raise TypeError
    return (name, domain)

# both should work
foo(name="a", domain="test")
foo(name="a", domain=["test"])


class TestFoo(unittest.TestCase):
    def test_foo(self): # raises an error
        self.assertRaises(TypeError, foo(name="a", domain=1))
    def test_foo2(self): # passes
        with self.assertRaises(TypeError):
            foo(name="a", domain=1)
    def test_foo3(self): # passes
        with pytest.raises(TypeError):
            foo(name="a", domain=1)

```

Here, `self.assertRaises(TypeError, foo(name="a", domain=1))` raises an error instead of failing because `foo` already raises a `TypeError` at runtime, which can't be caught out of context. We could enforce this, but it would contradict the docstring of `pybamm.Variable()`, which says that the argument is "domain : iterable of str", and changing the behaviour of the function means this is a breaking change – not recommended and outside the scope of migrating our testing infrastructure. I think we should be okay with removing this test case, since the behaviour is clearly correct – it shouldn't raise an error since we don't validate the type in the code. Do you have any thoughts, @Saransh-cpp?

Reply:
Totally agreed. We should remove this and the following test. Waiting on Saransh reply.

Reply:
Removing the test case sounds good to me too